mio/src/lib.rs

```diff
@@ -246,49 +307,66 @@ impl Loop {
                 }
             }
 
+            // TODO: attempt to coalesce events into token sets when they are
+            // for the same Wake
             for i in 0..amt {
                 let event = self.io.events().get(i).unwrap();
                 let token = event.token().as_usize();
                 if token == 0 {
                     while let Ok(msg) = self.rx.try_recv() {
                         self.notify(msg);
                     }
-                } else if let Some(complete) = self.done.remove(&token) {
-                    complete.finish(());
+                } else if let Some(wake) = self.dispatch.remove(&token) {
+                    wake.wake(&Tokens::from_usize(token));
                 }
             }
         }
     }
 
+    fn register_(&mut self, source: Source) {
+        self.io.register(&*source, mio::Token(0), mio::EventSet::none(), mio::PollOpt::empty());
+    }
+
+    fn schedule_(&mut self, token: usize, interest: Interest) {
+        let Interest { waiter, source, events, first_time } = interest;
+        let old = self.dispatch.insert(token, waiter);
+        debug_assert!(old.is_none());
+
+        // TODO handle failure
```

Comment:
I think the way we'll end up handling failure here will look like:

* See the error
* Store the error with the token somewhere
* Call the `waiter` callback

Communicating to the original future efficiently that it should check the storage may be... interesting. Just need to figure out how to get it off the fast path though, I don't think this will come up much in practice.